File: svm/zig/svm.zig
import std/math/linalg/vector;

struct SVM {
    weight: vec2,
    bias: f32,
}

impl SVM {
    fn new(c: f32, kernel: &str) -> SVM {
        SVM {
            weight: vec2(0.0, 0.0),
            bias: 0.0,
        }
    }

    fn fit(&mut self, x: []const vec2, y: []const i32, n: usize) {
        for _ in 0..n {
            let score = self.weight * x[0] + self.bias;
            if y[0] * score <= 1.0 {
                for i in 0..2 {
                    self.weight[i] += c * y[0] * x[0][i];
                }
                self.bias += c * y[0];
            }
        }
    }

    fn predict(&self, x: vec2) -> i32 {
        let score = self.weight * x + self.bias;
        if score >= 0.0 { 1 } else { -1 }
    }
}

pub fn main() {
    let svm = SVM::new(1.0, "linear");
    svm.fit(
        &[vec2(1.0, 2.0), vec2(3.0, 4.0)],
        &[1, -1],
        2,
    );
    assert(svm.predict(vec2(5.0, 6.0)) == 1);
}

File: svm/java/SVM.java
import java.util.Arrays;

public class SVM {

    private double C;
    private String kernel;
    private double[] w;
    private double b;

    public SVM(double C, String kernel) {
        this.C = C;
        this.kernel = kernel;
        this.w = new double[]{0.0, 0.0};
        this.b = 0.0;
    }

    public void fit(double[][] X, int[] y, int n) {
        for (int i = 0; i < n; i++) {
            double score = 0.0;
            for (int j = 0; j < 2; j++) {
                score += w[j] * X[i][j];
            }
            if (y[i] * score + b <= 1.0) {
                for (int j = 0; j < 2; j++) {
                    w[j] += C * y[i] * X[i][j];
                }
                b += C * y[i];
            }
        }
    }

    public int predict(double[] X) {
        double score = 0.0;
        for (int j = 0; j < 2; j++) {
            score += w[j] * X[j];
        }
        return score >= 0.0 ? 1 : -1;
    }

    public static void main(String[] args) {
        SVM svm = new SVM(1.0, "linear");
        svm.fit(new double[][]{{1, 2}, {3, 4}}, new int[]{1, -1}, 2);
        System.out.println(svm.predict(new double[]{5, 6}));
    }
}
